### PR TITLE
[core] Consolidate `@InternalApi` classes

### LINKED ISSUE #4348: [core] Consolidate `@InternalApi` classes
* Review, which classes are marked as internal api
* Move these classes into `internal` sub-packages, if possible. Or make this classes package-private.

## Annotated with @InternalApi

:heavy_check_mark: everything done now

-----
## Done
Everything that's done...

### pmd-compat6 - we keep this as is, no additional change in this module
* net.sourceforge.pmd.util.filter.Filters
* net.sourceforge.pmd.Report#Report
* net.sourceforge.pmd.Report#addRuleViolation
* net.sourceforge.pmd.Report#addConfigError
* net.sourceforge.pmd.Report#addError

### pmd-core
* now public API
  * net.sourceforge.pmd.lang.document.FileCollector
    * #getCollectedFiles and #getReporter are promoted to public API
* done - in some other way
  * net.sourceforge.pmd.renderers.AbstractAccumulatingRenderer#renderFileReport
    * the method is made final, so that it can't be overridden anymore in subclasses. That's how the AccumulatingRenderer works now.

* to be hidden
  * net.sourceforge.pmd.cache.AbstractAnalysisCache
  * net.sourceforge.pmd.cache.AnalysisCache
  * net.sourceforge.pmd.cache.AnalysisCacheListener
  * net.sourceforge.pmd.cache.AnalysisResult
  * net.sourceforge.pmd.cache.CachedRuleMapper
  * net.sourceforge.pmd.cache.CachedRuleViolation
  * net.sourceforge.pmd.cache.ChecksumAware
  * net.sourceforge.pmd.cache.FileAnalysisCache
  * net.sourceforge.pmd.cache.NoopAnalysisCache
  * whole package net.sourceforge.pmd.cache
  * net.sourceforge.pmd.lang.LanguageProcessor.AnalysisTask#AnalysisTask
    * also make net.sourceforge.pmd.lang.LanguageProcessor.AnalysisTask#withFiles internal, that's the other way to create a new AnalysisTask
    * Accessible through a InternalApiBridge
* net.sourceforge.pmd.util.log.MessageReporter (TODO: Rename to PmdReporter)
* Already in internal packages
  * net.sourceforge.pmd.util.log.internal.ErrorsAsWarningsReporter
  * net.sourceforge.pmd.util.log.internal.QuietReporter
  * net.sourceforge.pmd.util.log.internal.SimpleMessageReporter
* net.sourceforge.pmd.util.ResourceLoader
* net.sourceforge.pmd.cpd.Tokens#Tokens()
  * is hidden
* net.sourceforge.pmd.cpd.Tokens
  * is not anymore InternalApi, because it is used in public API: net.sourceforge.pmd.cpd.TokenEntry#getImage
* net.sourceforge.pmd.lang.rule.RuleTargetSelector#isRuleChain
  * removed, as it was only used in a unit test
* net.sourceforge.pmd.rules.RuleFactory (#4799 !!)
  * already made package private with the PR
* net.sourceforge.pmd.RuleSetReference (#4799 !!)
  * already moved into internal package with the PR
* net.sourceforge.pmd.RuleSetReferenceId (#4799 !!)
  * already moved into internal package with the PR
* net.sourceforge.pmd.RuleSets (#4799 !!)
  * already moved into internal package with the PR
* net.sourceforge.pmd.Report - made the following methods private
  * Report (ctor)
  * addRuleViolation
  * addConfigError
  * addError
  * Note: The class is moved with #4803 ...
* net.sourceforge.pmd.RuleContext
  * getRule() is package private
  * create(...) is moved to new InternalApiBridge
  * net.sourceforge.pmd.RuleContext#addViolationNoSuppress (not used?)
    * has been removed
    * See #2736 and #2896

* To be done after #4803 is merged:
  * net.sourceforge.pmd.lang.rule.ParametricRuleViolation (#4803 !!)
    * Maybe most of the constructors can be made package private. It's only created in RuleContext... this might need to stay internal API - maybe it can be moved into InternalApiBridge as well...
    * Should be made package private and added to the InternalApiBridge
* To be done after #4799 is merged:
  * net.sourceforge.pmd.RuleSet
    * applies(net.sourceforge.pmd.lang.document.FileId)
    * applies(net.sourceforge.pmd.Rule, net.sourceforge.pmd.lang.LanguageVersion)
    * These should be made package private and a InternalApiBridge class should be provided to make them accessible where needed.
  * net.sourceforge.pmd.RuleSetLoader#loadRuleSetsWithoutException
    * Should be made package private and added to the InternalApiBridge
  * net.sourceforge.pmd.RuleSetLoadException#RuleSetLoadException(net.sourceforge.pmd.RuleSetReferenceId, java.lang.Throwable)
    * Should be made package private and added to the InternalApiBridge
  * net.sourceforge.pmd.RuleSetLoadException#RuleSetLoadException(net.sourceforge.pmd.RuleSetReferenceId, java.lang.String)
    * Should be made package private and added to the InternalApiBridge
* net.sourceforge.pmd.RuleSet#removeDysfunctionalRules (#4793 !!)

* net.sourceforge.pmd.lang.ast.LexException#LexException(boolean, java.lang.String, int, int, java.lang.String, char) (previously TokenMgrError)
  * called only from generated TokenManagers
  * should be made package private and accessed via a InternalApiBridge
* net.sourceforge.pmd.PMDConfiguration#getAnalysisCache
  * make package private
* net.sourceforge.pmd.PMDConfiguration#setAnalysisCache(AnalysisCache)
  * make package private
  * Note: the other setAnalysisCache method, which takes a String, is public API
* net.sourceforge.pmd.lang.document.FileCollector#newCollector(net.sourceforge.pmd.lang.LanguageVersionDiscoverer, net.sourceforge.pmd.util.log.PmdReporter)
  * make package private and access via InternalApiBridge
* net.sourceforge.pmd.lang.document.FileCollector#newCollector(net.sourceforge.pmd.util.log.PmdReporter)
  * make package private and access via InternalApiBridge
* net.sourceforge.pmd.lang.rule.xpath.Attribute#replacementIfDeprecated
  * make package private and access via InternalApiBridge
* net.sourceforge.pmd.properties.PropertyDescriptor#getTypeId
  * make package private and access via InternalApiBridge
* net.sourceforge.pmd.properties.PropertyTypeId
  * move into internal package

### pmd-ant
* net.sourceforge.pmd.ant.Formatter
  * Note: the whole class can't be internal API, since this is part of the public ant api of PMD (see PMDTask). But the single methods annotated with InternalApi should be hidden
* net.sourceforge.pmd.ant.Formatter#getRenderer
* net.sourceforge.pmd.ant.Formatter#start
* net.sourceforge.pmd.ant.Formatter#end
* net.sourceforge.pmd.ant.Formatter#isNoOutputSupplied
* net.sourceforge.pmd.ant.Formatter#newListener
* net.sourceforge.pmd.ant.PMDTask#getRelativizeRoots

### pmd-apex
* net.sourceforge.pmd.lang.apex.ast.AbstractApexNode#getNode
  * should be package private
  * remove from ApexNode interface
* net.sourceforge.pmd.lang.apex.ast.ApexParser
  * should not be internal api, no reason why it should.
* net.sourceforge.pmd.lang.apex.multifile.ApexMultifileAnalysis#ApexMultifileAnalysis
  * should be package private. We need a InternalApiBridge to access it
* net.sourceforge.pmd.lang.apex.rule.design.AbstractNcssCountRule
  * should be package private
* net.sourceforge.pmd.lang.apex.rule.AbstractApexUnitTestRule
  * should be package private
* net.sourceforge.pmd.lang.apex.rule.internal.Helper (already internal)
  * no need for the annotation due to internal package
* Note: probably creates conflicts with #4806

### pmd-java
* net.sourceforge.pmd.lang.java.ast.InternalApiBridge
  * that's fine, nothing to change
* net.sourceforge.pmd.lang.java.rule.AbstractIgnoredAnnotationRule
  * move into internal package
* net.sourceforge.pmd.lang.java.symbols.JElementSymbol (experimental)
  * make it public. It is part of public API net.sourceforge.pmd.lang.java.symbols.SymbolVisitor
* net.sourceforge.pmd.lang.java.types.ast.LazyTypeResolver
  * move into internal package
* net.sourceforge.pmd.lang.java.types.JMethodSig#internalApi
  * the method should be removed. We should explicitly downcast to the internal api interface where needed.
* net.sourceforge.pmd.lang.java.types.TypeOps#isSameTypeInInference
  * make package private
  * and add a InternalApiBridge class

### pmd-jsp
* net.sourceforge.pmd.lang.jsp.ast.JspParser#getTokenBehavior
  * move into a InternalApiBridge
  * note: protected method tokenBehavior does the same...
* net.sourceforge.pmd.lang.jsp.ast.OpenTagRegister (is already package private)
  * no need for the annotation

### pmd-modelica
* net.sourceforge.pmd.lang.modelica.ast.InternalModelicaNodeApi
  * rename to InternalApiBridge to be consistent
* net.sourceforge.pmd.lang.modelica.resolver.InternalModelicaResolverApi
  * rename to InternalApiBridge to be consistent
* net.sourceforge.pmd.lang.modelica.resolver.ModelicaSymbolFacade
* net.sourceforge.pmd.lang.modelica.resolver.ResolutionContext

### pmd-plsql
* net.sourceforge.pmd.lang.plsql.ast.AbstractSelectStatement (is already package private)
* net.sourceforge.pmd.lang.plsql.ast.InternalApiBridge
* net.sourceforge.pmd.lang.plsql.rule.design.AbstractNcssCountRule

### pmd-scala
* net.sourceforge.pmd.lang.scala.ScalaLanguageModule#dialectOf

### pmd-visualforce
* net.sourceforge.pmd.lang.vf.ast.OpenTagRegister
* net.sourceforge.pmd.lang.vf.ast.VfAstInternals

### doc
* Accept https://docs.pmd-code.org/snapshot/pmd_projectdocs_decisions_adr_3.html#separation-between-public-api-internal-and-implementation
  * I guess, I forgot to update this when I merged #4756

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

* [x] [core] Internalize AnalysisCache and related types
  *  net.sourceforge.pmd.cache.AbstractAnalysisCache
  *  net.sourceforge.pmd.cache.AnalysisCache
  *  net.sourceforge.pmd.cache.AnalysisCacheListener
  *  net.sourceforge.pmd.cache.AnalysisResult
  *  net.sourceforge.pmd.cache.CachedRuleMapper
  *  net.sourceforge.pmd.cache.CachedRuleViolation
  *  net.sourceforge.pmd.cache.ChecksumAware
  *  net.sourceforge.pmd.cache.FileAnalysisCache
  *  net.sourceforge.pmd.cache.NoopAnalysisCache
  *  whole package net.sourceforge.pmd.cache
* [x] PmdReporter
  * Already in internal packages
    * net.sourceforge.pmd.util.log.internal.ErrorsAsWarningsReporter
    * net.sourceforge.pmd.util.log.internal.QuietReporter
    * net.sourceforge.pmd.util.log.internal.SimpleMessageReporter
  * Renamed MessageReporter to PmdReporter - and removed InternalApi, since the reporter is supposed to be configurable via AbstractConfiguration#setReporter, so is part of the public API
* [x] net.sourceforge.pmd.util.ResourceLoader
  * moved to net.sourceforge.pmd.util.internal.ResourceLoader
* [x] net.sourceforge.pmd.cpd.Tokens
  * ctor net.sourceforge.pmd.cpd.Tokens#Tokens() is hidden now
  * the class itself is not anymore InternalApi, because it is used in public API: net.sourceforge.pmd.cpd.TokenEntry#getImage
* [x] net.sourceforge.pmd.properties.PropertyTypeId
* [x] net.sourceforge.pmd.properties.PropertyDescriptor#getTypeId
* [x] net.sourceforge.pmd.lang.document.FileCollector
  * #newCollector(net.sourceforge.pmd.lang.LanguageVersionDiscoverer, net.sourceforge.pmd.util.log.MessageReporter) and #newCollector(net.sourceforge.pmd.util.log.MessageReporter)
  * #getCollectedFiles and #getReporter are promoted to public API
* [x] net.sourceforge.pmd.lang.rule.RuleTargetSelector#isRuleChain
  * removed, as it was only used in a unit test
* [x] net.sourceforge.pmd.lang.LanguageProcessor.AnalysisTask#AnalysisTask
  * also make net.sourceforge.pmd.lang.LanguageProcessor.AnalysisTask#withFiles internal, that's the other way to create a new AnalysisTask
* [x] net.sourceforge.pmd.renderers.AbstractAccumulatingRenderer#renderFileReport
  * the method is made final, so that it can't be overridden anymore in subclasses. That's how the AccumulatingRenderer works now.
* [x] net.sourceforge.pmd.Report - made the following methods private
  * Report (ctor)
  * addRuleViolation
  * addConfigError
  * addError
  * Note: The class is moved with #4803 ...
* [x] net.sourceforge.pmd.RuleContext
  * getRule() is package private
  * create(...) is moved to new InternalApiBridge
  * net.sourceforge.pmd.RuleContext#addViolationNoSuppress (not used?)
    * has been removed
    * See #2736 and #2896
  * Note: The new InternalApiBridge class needs to be moved into package reporting after #4803  ...

* [x] pmd-ant
  * Formatter
    * The method `getRenderer()` has been removed (not used, can access field directly here)
    * The method `start(String)` is private now and not accessible anymore.
    * The method `end(Report)` has been removed (has not been used)
    * The method `isNoOutputSupplied()` is now package private and not accessible anymore.
    * The method `newListener(Project)` is now package private and not accessible anymore. Added InternalApiBridge, so that we access it from PMDTaskImpl
  * PMDTask
    * The method `getRelativizeRoots()` has been removed - actually moved into PMDTaskImpl

* [x] pmd-apex - cleanups done, see #4348

* [x] pmd-java - cleanups done, see #4348

* [x] pmd-jsp
* [x] pmd-modelica
* [x] pmd-plsql
* [x] pmd-scala
* [x] pmd-visualforce
* [x] Accept ADR 3 (#4756)
* [x] javadoc generation - exclude internal

* [x] net.sourceforge.pmd.lang.rule.ParametricRuleViolation - to be done after #4803 
* [x] net.sourceforge.pmd.RuleSet etc. - to be done after #4799 

## Related issues

- Fixes #4348

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/7671512195">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-ant/src/main/java/net/sourceforge/pmd/ant/Formatter.java [hunk 1]
```diff
@@ -37,7 +35,6 @@
 import net.sourceforge.pmd.reporting.GlobalAnalysisListener;
 import net.sourceforge.pmd.reporting.ListenerInitializer;
 
-@InternalApi
 public class Formatter {
```
Comment (oowekyala) [inline pmd-ant/src/main/java/net/sourceforge/pmd/ant/Formatter.java:38]: Can we move that to the internal package? That way we wouldn't need the internal API bridge class

**Review by oowekyala — COMMENTED** [review]
Thanks! We should also exclude internal packages from the javadoc generation. I remember there was a problem with using regex for that, but maybe we can just hardcode the internal packages in the pom.

File: pmd-ant/src/main/java/net/sourceforge/pmd/ant/Formatter.java (hunk 1 above)
Comment (adangel) [reply pmd-ant/src/main/java/net/sourceforge/pmd/ant/Formatter.java:38]: I consider this class as public API, as this is used to configure the PMD Task:
e.g. https://docs.pmd-code.org/apidocs/pmd-ant/7.0.0-SNAPSHOT/net/sourceforge/pmd/ant/PMDTask.html#addFormatter(net.sourceforge.pmd.ant.Formatter)

This is basically, what ant calls, when it sees this in xml:

```xml
<pmd ...>
    <formatter type="html" toFile="${build}/pmd_report.html"/>
</pmd>
```

Well, it creates a new instance of Formatter, calls [setType](https://docs.pmd-code.org/apidocs/pmd-ant/7.0.0-SNAPSHOT/net/sourceforge/pmd/ant/Formatter.html#setType(java.lang.String)) on it and [setToFile](https://docs.pmd-code.org/apidocs/pmd-ant/7.0.0-SNAPSHOT/net/sourceforge/pmd/ant/Formatter.html#setToFile(java.io.File)), and then calls `addFormatter` on the PMDTask instance.

I think, the Formatter should really only be a data class without any logic (it's part of the configuration of the ant task). Right now, it seems to do a bit more (like creating a renderer by itself - which should really be part of the implementation PMDTaskImpl...). But that's something to refactor later.

Having these InternalApiBridge classes is kind of a compromise I guess. Not nice, but does the job of hiding methods in public classes.

**Comment by adangel** [comment]
> We should also exclude internal packages from the javadoc generation. 

Thanks for the reminder. I've updated the pattern to include one or two possible subpackages. As we defined the pattern with a wildcard, the subpackages weren't included (I guess, if we add each internal package individually, the subpackages would be automatically excluded as well, but then we need to update the javadoc config everytime we add a internal package).

File: pmd-ant/src/main/java/net/sourceforge/pmd/ant/Formatter.java (hunk 1 above)
Comment (oowekyala) [reply pmd-ant/src/main/java/net/sourceforge/pmd/ant/Formatter.java:38]: Ok thanks for explaining, that makes sense. Probably this should be mentioned in a comment in the code

**Comment by adangel** [comment]
This is finally ready now.

**Pull request merged**